rlclientlib/federation/vw_local_joiner.cc

```diff
@@ -0,0 +1,222 @@
+      // TODO: real timestamp
+      messages::flatbuff::v2::TimeStamp ts(2020, 3, 1, 10, 20, 30, 0);
+      const auto* event_payload = message->payload();
+      auto vec = fbb.CreateVector(event_payload->data(), event_payload->size());
+      auto fb = messages::flatbuff::v2::CreateJoinedEvent(fbb, vec, &ts);
+      fbb.Finish(fb);
+      auto detached_buffer = fbb.Release();
+
+      const auto* je = flatbuffers::GetRoot<messages::flatbuff::v2::JoinedEvent>(detached_buffer.data());
+      _joiner->process_event(*je);
+    }
+  }
+
+  return error_code::success;
+}
+
+int vw_local_joiner::invoke_join(std::unique_ptr<i_joined_log_batch>& batch_out, api_status* status)
+{
+  std::stringstream ss;
+  ss << "Joining " << _joiner->events_in_queue() << " events." << std::endl;
+  TRACE_INFO(_trace_logger, ss.str());
+
+  std::vector<VW::example*> examples;
+  std::unique_ptr<vw_joined_log_batch> joined_batch = VW::make_unique<vw_joined_log_batch>(_joiner_workspace);
+
+  while (_joiner->processing_batch())
+  {
+    examples.push_back(VW::new_unused_example(*_joiner_workspace));
+
+    // False means there was a problem and we should try reading the next one.
+    if (!_joiner->process_joined(examples))
+    {
+      assert(examples.size() == 1);
+      auto* ex_to_return = examples.back();
+      examples.pop_back();
+      VW::finish_example(*_joiner_workspace, *ex_to_return);
+      continue;
+    }
+    VW::setup_examples(*_joiner_workspace, examples);
+
+    // We must remove the trailing newline example.
+    auto* newline_ex = examples.back();
+    examples.pop_back();
+    VW::finish_example(*_joiner_workspace, *newline_ex);
+
+    // Add examples to batch
+    for (VW::example* e : examples) { joined_batch->add_example(e); }
+    examples.clear();
+  }
+
+  batch_out.reset(joined_batch.release());
+  return error_code::success;
+}
+
+vw_joined_log_batch::vw_joined_log_batch(std::shared_ptr<VW::workspace> joiner_workspace)
+    : _joiner_workspace(std::move(joiner_workspace))
+{
+}
+
+vw_joined_log_batch::~vw_joined_log_batch()
+{
+  for (auto* example : _examples) { _joiner_workspace->finish_example(*example); }
+
+  if (_output_example_ptr != nullptr)
+  {
+    _joiner_workspace->finish_example(*_output_example_ptr);
+    _output_example_ptr = nullptr;
+  }
+}
+
+void vw_joined_log_batch::add_example(VW::example* example) { _examples.push_back(example); }
+
+int vw_joined_log_batch::next(std::unique_ptr<VW::io::reader>& chunk_reader, api_status* status)
+{
+  // free the previous output example, if it exists
+  if (_output_example_ptr != nullptr)
+  {
+    _joiner_workspace->finish_example(*_output_example_ptr);
+    _output_example_ptr = nullptr;
+  }
+
+  // clear output state
+  chunk_reader.reset(nullptr);
+  _output_example_buffer = std::make_shared<std::vector<char>>();
+
+  // output the next example in the batch
+  if (!_examples.empty())
+  {
+    io_buf io_writer;
+    VW::details::cache_temp_buffer temp_buffer;
```

Comment:
Rather than using cache format I had it my head that this reader would expose binary log based examples that had been EUD joined so that the consumer of this reader could just make the output available as the input to a driver